internal/core/src/segcore/reduce_c.cpp

```diff
@@ -154,6 +156,208 @@ ReduceResultData(std::vector<SearchResult*>& search_results, int64_t nq, int64_t
+        auto result_distances = std::vector<float>(nq * topK);
+
+        // init result output fields data
+        auto& output_fields_meta = search_results[0]->output_fields_meta_;
+        auto num_output_fields = output_fields_meta.size();
+        auto result_output_fields_data = std::vector<milvus::aligned_vector<char>>(num_output_fields);
+        for (int i = 0; i < num_output_fields; i++) {
+            auto size = output_fields_meta[i].get_sizeof();
+            result_output_fields_data[i].resize(size * nq * topK);
+        }
+
+        // Reorganize search results, get result ids, distances and output fields data
+        ReorganizeSearchResultsV2(search_results, nq, topK, result_ids, result_distances, result_output_fields_data);
+
+        // prefix sum, get slices offsets
+        AssertInfo(num_slices > 0, "empty nq_slice_sizes is not allowed");
+        auto slice_offsets_size = num_slices + 1;
+        auto slice_offsets = std::vector<int32_t>(slice_offsets_size);
+        slice_offsets[0] = 0;
+        slice_offsets[1] = nq_slice_sizes[0];
+        for (int i = 2; i < slice_offsets_size; i++) {
+            slice_offsets[i] = slice_offsets[i - 1] + nq_slice_sizes[i - 1];
+        }
+        AssertInfo(slice_offsets[num_slices] == nq,
+                   "illegal req sizes"
+                   ", slice_offsets[last] = " +
+                       std::to_string(slice_offsets[num_slices]) + ", nq = " + std::to_string(nq));
+
+        // get search result data blobs by slices
+        auto blobs = static_cast<CProto*>(malloc(sizeof(CProto) * num_slices));
+#pragma omp parallel for
+        for (int i = 0; i < num_slices; i++) {
+            auto cProto = GetSearchResultDataSlice(result_ids, result_distances, result_output_fields_data, nq, topK,
+                                                   slice_offsets[i], slice_offsets[i + 1], output_fields_meta);
+            blobs[i] = cProto;
+        }
+
+        // set final result ptr
+        cSearchResultDataBlobs->num_blobs = num_slices;
+        cSearchResultDataBlobs->blobs = blobs;
+        return milvus::SuccessCStatus();
+    } catch (std::exception& e) {
+        return milvus::FailureCStatus(UnexpectedError, e.what());
+    }
+}
+
+CStatus
+GetSearchResultDataBlob(CProto* searchResultDataBlob,
+                        CSearchResultDataBlobs* cSearchResultDataBlobs,
+                        int32_t blob_index) {
+    try {
+        AssertInfo(blob_index < cSearchResultDataBlobs->num_blobs, "blob_index out of range");
+        searchResultDataBlob->proto_blob = cSearchResultDataBlobs->blobs[blob_index].proto_blob;
+        searchResultDataBlob->proto_size = cSearchResultDataBlobs->blobs[blob_index].proto_size;
+        return milvus::SuccessCStatus();
+    } catch (std::exception& e) {
+        return milvus::FailureCStatus(UnexpectedError, e.what());
+    }
+}
+
+void
+DeleteSearchResultDataBlobs(CSearchResultDataBlobs* cSearchResultDataBlobs) {
```

Comment:
Write custom destructors for struct cSearchResultDataBlobs!

Reply:
fixed

Reply:
cgo doesn't support destructor, which is a C++ feature

Reply:
wrap SearchResultDataBlobs，and typedef void* CSearchResultDataBlobs